File: src/library/java/gg/generations/rarecandy/pokeutils/IHideDuringAnimation.java
package gg.generations.rarecandy.pokeutils;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import gg.generations.rarecandy.pokeutils.util.JsonUtils;

import java.util.Collections;
import java.util.List;

public interface IHideDuringAnimation {
    IHideDuringAnimation NONE = new IHideDuringAnimation() {
        @Override
        public boolean blackList() {
            return false;
        }

        @Override
        public List<String> animations() {
            return Collections.emptyList();
        }
    };

    boolean blackList();
    List<String> animations();

    default boolean check(String animation) {
        return animations().contains(animation) != blackList();
    }

    static IHideDuringAnimation deserialize(JsonElement element) {
        var obj = element.getAsJsonObject();
        var blackList = JsonUtils.extractIfPresent(obj, "blackList", false, JsonElement::getAsBoolean);
        var animations = JsonUtils.extractListIfPresent(obj, "animations", JsonElement::getAsString);

        return IModelConfig.Factory.ACTIVE_FACTORY.createHideDuringAnimation(blackList, animations);
    }

    static JsonElement serialize(IHideDuringAnimation hideDuringAnimation) {
        var obj = new JsonObject();
        JsonUtils.putIf(value -> value, obj, "blackList", hideDuringAnimation.blackList(), JsonPrimitive::new);
        JsonUtils.putListIf(obj, "animations", hideDuringAnimation.animations(), JsonPrimitive::new);
        return obj;
    }

    //        public boolean check(Animation animation) {
//            return check(animation != null ? animation.name : null);
//        }
}

File: src/library/java/gg/generations/rarecandy/pokeutils/IHideDuringAnimation.kt
package gg.generations.rarecandy.pokeutils

import gg.generations.rarecandy.codec.BOOL
import gg.generations.rarecandy.codec.STRING_LIST
import gg.generations.rarecandy.codec.codec
import gg.generations.rarecandy.codec.obj
import gg.generations.rarecandy.codec.putIfNotEquals
import gg.generations.rarecandy.codec.read

data class IHideDuringAnimation(
    val blackList: Boolean,
    val animations: MutableList<String>
) {



    fun check(animation: String?): Boolean {
        return animations.contains(animation) != blackList
    }

    companion object {
        val CODEC = codec({ hide ->
            obj().also {
                it.putIfNotEquals("blackList", BOOL, { !it }, hide.blackList)
                it.putIfNotEquals("animations", STRING_LIST, { it.isEmpty() }, hide.animations)
            }
        }, {
            it.obj().let {
                IHideDuringAnimation(
                    it.read("blackList", BOOL, false),
                    it.read("animations", STRING_LIST, mutableListOf())
                )
            }
        })

        val NONE: IHideDuringAnimation = IHideDuringAnimation(false, mutableListOf())
    }
}
